### PR TITLE
Adjust io-threads number in runtime to fully utilize multi threads and make CPU more efficient.

### PR DESCRIPTION
**Pull request by zhulipeng**
## Description
This patch try to fix the issue when io-threads number is configured, io-threads will be stopped according below logic, this will cause the io-threads useless and longer latency.  According to existing logic, the io-threads will be activated by a hardcode threshold value,  this will cause QPS regression by the different configured io-threads number even the same clients requests.
```c
int stopThreadedIOIfNeeded(void) {
    int pending = listLength(server.clients_pending_write);

    /* Return ASAP if IO threads are disabled (single threaded mode). */
    if (server.io_threads_num == 1) return 1;

    if (pending < (server.io_threads_num*2)) {
        if (server.io_threads_active) stopThreadedIO();
        return 1;
    } else {
        return 0;
    }
}
```
With this patch, it will dynamically adjust the /IO threads number in runtime according to real workloads to tradeoff between the latency and CPU efficiency. I use decay rate formula based on `clients_pending_writes` to make throughput stable. 

### Test Environment

- OS: CentOS Stream 8
- Kernel: 6.2.0
- Platform: Intel Xeon Platinum 8380
- Base commit: b2a397366b1704b03f5042395ff620cd2096f598

#### Start Valkey-Server
Start valkey-server `taskset -c 0-3 ~/src/valkey-server /tmp/valkey_1.conf` with below config.

```
port 9001
bind * -::*
daemonize yes
protected-mode no
save ""
io-threads 4
io-threads-do-reads yes
```
#### Test from throughput perspective 
Start valkey-benchmark in localhost with below commands:
`taskset -c 4-7 ./src/valkey-benchmark -p 9001 -t set,get -d 128 -r 5000000 -n 10000000 --threads 2 -c 10` 
to measure the **SET** **GET** performance gap correspondingly, results are listed as below: 
we can find that the io-threads is not used due to `clients_pending_write < (server.io_threads_num*2)`, the throughput is limited by only single thread even more CPU resources were allocated. 

**NOTE: QPS will use relative value instead of absolute value.**
**CPU utilized:**  ```perf stat -p `pidof valkey-server` sleep 5```
||Base CPU utilized|Opt CPU utilized|QPS Gain(Opt/Base-1)|
|-|-|-|-|
|SET|1|2|23%|
|GET|1|2|25%|

#### Test from CPU efficiency perspective
If we increase the requests number by below commands to make origin io-threads work, throughput increased as expected, but all CPU are fully utilized. With this patch, we can make CPU more efficiency by reducing the active io-threads number.
`taskset -c 4-7 ./src/valkey-benchmark -p 9001 -t set,get -d 128 -r 5000000 -n 10000000 --threads 2 -c 15` 

||Base CPU utilized|Opt CPU utilized|QPS Gain(Opt/Base-1)|
|-|-|-|-|
|SET|3.999|2.908|8%|
|GET|3.999|2.918|3%|

One test case issue is related to https://github.com/valkey-io/valkey/issues/397

### CODE REVIEW

**Comment by zhulipeng** [comment]
@valkey-io/core-team Could you help to review this patch?

**Comment by madolson** [comment]
@lipzhu I noticed there wasn't a test, we should at least add one for that.

**Comment by zhulipeng** [comment]
@madolson I am thinking how to add the unit tests for io-threads, it's for performance purpose, I am uncertain how to simulate it in unit tests, can you provide some guidance?

**Comment by zhulipeng** [comment]
Ping @madolson .

File: src/networking.c
Comment (PingXie) [inline src/networking.c]:
```suggestion
    /* We start from the main thread, no additional io-threads created yet. */
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]:
```suggestion
/* This function dynamically adjusts the I/O thread number according to
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: I think a hard-coded smoothing factor of 0.8 makes sense. It decays the moving average by an order of magnitude every 10 rounds, with every round lasting about 100 ms by default.

consider giving 0.8 a symbolic name and computing 0.2 from 0.8 too?

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: I think the readability can be improved a bit here. consider the following?

```suggestion
    if (server.io_threads_active_num < target_num) {
        /* Increasing active threads */
        for (size_t i = server.io_threads_active_num; i < target_num; ++i) {
            pthread_mutex_unlock(&io_threads_mutex[i]);
        }
    } else {
        /* Decreasing active threads */
        for (size_t i = server.io_threads_active_num; i > target_num; --i) {
            pthread_mutex_lock(&io_threads_mutex[i - 1]);
        }
    }
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: With the function name change, the return value (of 0 enabling IO threading and 1 disabling IO threading) is confusing. I think the old name is fine (and clearer).

File: tests/integration/io-threads.tcl
Comment (PingXie) [inline tests/integration/io-threads.tcl:31]: Can we add a test case that validates 

1. the active thread count can increase to the max
2. the active thread count can decay back to 1

Writing a reliable test for this feature is tricky and we might need to introduce additional `DEBUG` hooks but I think it will be worth the effort.

**Comment by PingXie** [comment]
My understanding of this change is that it helps with the volatile workload case where request rates fluctuate quite a bit. This change doesn't allocate/activate more IO worker threads at runtime but it allows the IO worker threads to stay active a bit longer using the Exponential Moving Average (EMA). This trades off a bit of CPU efficiency for a more consistent latency through these workload spikes/troughs, potentially significant. 

Overall I am onboard with this change. I think the only thing I am looking forward to is a test that proves we can still activate all the threads configured via `io-threads` and the IO threads can indeed quiesce.

**Comment by PingXie** [comment]
> I am thinking how to add the unit tests for io-threads, it's for performance purpose, I am uncertain how to simulate it in unit tests, can you provide some guidance?

I don't think we need a (unit) test for performance. We will use release candidates to get feedback. What we really need is a correctness test as I mentioned above.

File: tests/integration/io-threads.tcl
Comment (zhulipeng) [reply tests/integration/io-threads.tcl:31]: > Writing a reliable test for this feature is tricky and we might need to introduce additional DEBUG hooks but I think it will be worth the effort

Can you provide more guidance of how to introduce additional DEBUG hooks? In my local, I used `memtier_benchmark -c 25 -t 4`  to simulate the multiple clients/requests, I am not sure if the TCL test framework had the similar API or where can I test the correctness, debug print a message of the `io_threads_active_num`?

File: src/networking.c
Comment (zhulipeng) [reply src/networking.c]: Thanks, changed.

File: src/networking.c
Comment (zhulipeng) [reply src/networking.c]: Revert name from `adjustThreadedIOIfNeeded` to `stopThreadedIOIfNeeded`

File: tests/integration/io-threads.tcl
Comment (zhulipeng) [reply tests/integration/io-threads.tcl:31]: Kindly ping @PingXie

File: src/server.h
Comment (PingXie) [inline src/server.h]:
```suggestion
    int io_threads_active_num;  /* Number of active IO threads to use. */
```

File: src/server.c
Comment (PingXie) [inline src/server.c:5405]: we should keep this line for compat.

```suggestion
            "io_threads_active:%i\r\n", server.io_threads_active_num > 1));
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]:
```suggestion
 * real workloads at runtime. Currently we track the latest x rounds of
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]:
```suggestion
 * pending writes as a measure of clients we need to handle in parallel.
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]:
```suggestion
 * However, the I/O threading is disabled globally for reads as well if
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: Hmm, this change would slow down the "scale up" operation as well (in addition to "scale down"), which would lead to a lower QPS on the rising edge. Consider this?

```suggestion
    int pending = listLength(server.clients_pending_write);
    server.clients_pending_write_avg_num = (int)max((
        server.clients_pending_write_avg_num * SMOOTH_FACTOR +
        pending * (1 - SMOOTH_FACTOR), pending);

```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: let's use the snake case style.

```suggestion
    int target_num = min(server.io_threads_num,
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: I think a cleaner way is to update `server.io_threads_active_num` to `target_num` before returning on L4359

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: Sorry for going back and forth on this one but on a second thought, I think your previous function name makes more sense. 

```suggestion
int adjustIOThreadCount(void) {
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]:
```suggestion
 * The function returns the active IO worker thread count. */
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: There are two options here: either return `server.io_threads_active_num` or return `void`. My preference would be returning the active IO thread count and this will remove all the references to the global var `server.io_threads_active_num` later on. I am always for less coupling, at any level.

```suggestion
    return server.io_threads_active_num;
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c:4382]:
```c
io_threads_active_num = adjustIOThreadCount();
if (io_threads_active_num == 1) {
```

File: src/server.c
Comment (PingXie) [inline src/server.c]:
```suggestion
    adjustIOThreadCount();
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c:4493]: This is not related to your change but the original logic doesn't seem right to me. With this logic the activation of io-threads-do-read implicitly depends on the activation of io-threads-do-write, which should've been two orthogonal decisions IMO. We can address this in a different PR.  
```suggestion
    if (!server.io_threads_do_reads) return 0;
    adjustIOThreadCount();
    if (server.io_threads_active_num == 1) return 0;
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: it will be ideal to reduce the dependency on global vars such as `server.io_threads_active_num`. since `adjustIOThreadCount()` already returns the number of active thread count, we could just use its return value. 

```suggestion
        int target_id = item_id % io_threads_active_num;
```

File: tests/integration/io-threads.tcl
Comment (PingXie) [reply tests/integration/io-threads.tcl:31]: I was thinking about adding some new `DBBUG` command to modify the engine behavior (it looks like you made a change in debug.c but it doesn't look like you've completed it). After seeing #344, I think a real unit test would be more preferable.

Comment (zhulipeng) [reply tests/integration/io-threads.tcl:31]: @PingXie Thanks for your effort to review this. 

> I was thinking about adding some new DBBUG command to modify the engine behavior (it looks like you made a change in debug.c but it doesn't look like you've completed it)

I just add a monitor for `DEBUG io-threads` command, sorry, I didn't get your point to modify the engine behavior, can you be more specific? 

> After seeing https://github.com/valkey-io/valkey/pull/344, I think a real unit test would be more preferable.

Do you want to add the unit test in this PR, or we can add the unit test when new test framework merged?

File: src/networking.c
Comment (zhulipeng) [reply src/networking.c:4493]: Sure, I will keep the logic as origin, and open another issue to address this.

File: tests/integration/io-threads.tcl
Comment (PingXie) [reply tests/integration/io-threads.tcl:31]: >can you be more specific?

No worries. I was thinking of manipulating the data structure like `clients_pending_write` in a new `DEBUG` command to simulate different workload. Now that we will have a true unit test framework, I don't think we should bother with this hack at all. Do you mind reverting the changes in debug.c? 

>Do you want to add the unit test in this PR, or we can add the unit test when new test framework merged?

I think @madolson is actively working on #344 and it is also close to completion. That said, I am fine with adding the tests in a separate PR and if you decide so, could you open an issue and assign it to yourself?

Comment (murphyjacob4) [reply tests/integration/io-threads.tcl:31]: I like the idea of adding some DEBUG commands so we can properly do TCL tests of IO threading features. It probably doesn't make sense to play with the pending writes directly, but something like `DEBUG set-active-io-thread-count` could be a good DEBUG command, even outside of just TCL testing, so that we can force IO threads on and off. E.g. I remember debugging some TLS related issues with IO threading that only happened when IO threads turned off, and it required me to run and stop valkey-benchmark repeatedly.

So I would suggest:

1. Unit testing the "pending writes" to "active IO threads" logic
2. E2E testing the whole IO thread flow with a DEBUG command that sends commands while adjusting IO thread count up and down

But we can follow up on those improvements

WDYT?

Comment (PingXie) [reply tests/integration/io-threads.tcl:31]: That is a good point. I think it would be good to address it separately though. Let's open a separate issue?

For this change, I am mostly interested in proving that the EMA logic is working properly (it can scale up and also scale back down) so I think a real unit test would be a lot less flaky to maintain in the long run.

Comment (zhulipeng) [reply tests/integration/io-threads.tcl:31]: @PingXie https://github.com/valkey-io/valkey/issues/396 https://github.com/valkey-io/valkey/issues/397 were created to track the debug commands and tests. @murphyjacob4 please feel free to add. BTW, I have no permission to assign the issues.

**Comment by zhulipeng** [comment]
@PingXie I added integration test to validate the io_threads_active_num EMA logic working properly. Could you help to review again?

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: Nit

```suggestion
/* This function dynamically adjusts the I/O thread number to match real-time
 * workloads. It tracks the latest 'x' rounds of pending writes as a measure of
 * clients needing parallel handling. If there are too few pending clients, I/O
 * threading is disabled globally, for both reads and writes. 
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]: Nit
```suggestion
    /* The target io-thread number should be within the range of [1, io_threads_num]. */
```

File: src/networking.c
Comment (PingXie) [inline src/networking.c]:
```suggestion
        /* We may still have clients with pending reads when this function
         * is called. Handle them before stopping the threads. */
```

File: src/server.c
Comment (PingXie) [inline src/server.c]:
```suggestion
    /* Adjust the I/O threads to match the real-time workloads. */
```

File: tests/integration/io-threads.tcl
Comment (PingXie) [inline tests/integration/io-threads.tcl]:
```suggestion
            set io_threads_active [get_info_field $server_info io_threads_active]
            set io_threads_maximum_num [get_info_field $server_info io_threads_maximum_num]
            set io_threads_active_num [get_info_field $server_info io_threads_active_num]
```

File: tests/integration/io-threads.tcl
Comment (PingXie) [inline tests/integration/io-threads.tcl:57]: This is quite a loop but judging by your test runs, this test doesn't seem to increase the test execution time.

Another concern of mine is the potential flakiness of this test since it has non-deterministic dependency in nature. 

That said, I am fine with keeping this test for now. Push comes shove, we can remove it since we will get coverage from #397 as well.

**Review by PingXie — COMMENTED** [review]
LTGM. Some final touches and this PR is good to go.

File: tests/integration/io-threads.tcl
Comment (zhulipeng) [reply tests/integration/io-threads.tcl]: Thanks, `proc get_info_field` is a common proc in cluster module, I just moved it to `support/util.tcl`.

File: tests/integration/io-threads.tcl
Comment (zhulipeng) [reply tests/integration/io-threads.tcl:57]: > this test doesn't seem to increase the test execution time

Depends on the write speed,  the condition of meet total count of 10000 should be very fast.

>Another concern of mine is the potential flakiness of this test since it has non-deterministic dependency in nature.
>
>That said, I am fine with keeping this test for now. Push comes shove, we can remove it since we will get coverage from https://github.com/valkey-io/valkey/issues/397 as well.

Got it.

**Review by PingXie — APPROVED** [review]
Thanks @lipzhu!

**Comment by zhulipeng** [comment]
What is the next process, is it ready to merge? @PingXie

File: src/server.c
Comment (madolson) [inline src/server.c]:
```suggestion
```
This is just a config value, I'm nots sure we need to print this out in info.

File: src/networking.c
Comment (madolson) [inline src/networking.c:4337]: How was this number determined? Is it just a number?

**Review by madolson — COMMENTED** [review]
LGTM , just two minor comments.

File: src/networking.c
Comment (zhulipeng) [reply src/networking.c:4337]: Yes, a hardcode number. It can take react to the target real time value, just as Ping mentioned, it decays the moving average by an order of magnitude every 10 rounds.

image

File: src/server.c
Comment (zhulipeng) [reply src/server.c]: @madolson Removed as suggested.

**Review by madolson — APPROVED** [review]

**Comment by madolson** [comment]
@valkey-io/core-team Need approval for adding a single new metric, `io_threads_active_num`, which will indicate the number of IO threads that are actively doing work. Some of the work in this PR overlaps with the other performance work that is ongoing, but the tests should all largely still be applicable. Asking for an async vote for the new metric.

The real world usage of the metric is probably so an operator can understand how many of the CPUs are actually doing useful work when debugging.

**Comment by zuiderkwast** [comment]
For some context, we already have `io_threads_active` 0 or 1 (on/off). I suppose we need to keep this one as is...

The config `io-threads` includes the main thread in the count, so the minimum is 1. Will `io_threads_active_num` include the main thread too?

**Comment by zhulipeng** [comment]
> For some context, we already have `io_threads_active` 0 or 1 (on/off). I suppose we need to keep this one as is...
> 

The `io_threads_active` is kept. 

> The config `io-threads` includes the main thread in the count, so the minimum is 1. Will `io_threads_active_num` include the main thread too?

Yes, `io_threads_active_num` started from 1, main thread is included.

**Review by hwware — APPROVED** [review]

File: src/networking.c
Comment (zuiderkwast) [inline src/networking.c:4333]: These changes will just cause merge conflicts for the Async IO threads feature. It's better that we don't merge it now.

Comment (zhulipeng) [reply src/networking.c:4333]: I saw the Async IO threads remove the `startThreadIO` and use the dynamic threads number too, this is similar with this patch. The difference is that Async IO threads keep the `threshold` based on the real time events count, this patch use  EMA to make a more consistent latency when requests spikes/troughs. 

Maybe we could merge this firstly and then rebase Async IO threads on this? If needed, I could help the rebase work.

Comment (zuiderkwast) [reply src/networking.c:4333]: @uriyage WDYT?

**Comment by zhulipeng** [comment]
So we are still waiting for the approval from @soloestoy @enjoy-binbin , could you help to take a look at this when you are free?

**Review by enjoy-binbin — APPROVED** [review]
LGTM, did not review the tests

**Comment by zuiderkwast** [comment]
I think it's better to merge Async IO threads first and then add this one afterwards, since this one is a smaller change.

**Comment by madolson** [comment]
@lipzhu Do you want to update this PR to handle the new IO threads or shall we close this?

**Comment by zhulipeng** [comment]
> @lipzhu Do you want to update this PR to handle the new IO threads or shall we close this?

Sure, let's close this PR as most logic is already covered by new IO threads. 
The difference is the EMA logic for `target thread num`, I can open a new PR in future if needed.

**Pull request closed without merging**
